src/game.cpp

```diff
@@ -11602,14 +11602,83 @@ void game::read()
     auto filter = []( const item &it ) {
         return it.is_book();
     };
-    int pos = inv_for_filter( _("Read:"), filter );
 
-    if (pos == INT_MIN) {
-        add_msg(_("Never mind."));
+    auto item_loc = inv_map_splice( filter, _("Read:") );
+
+    const int inv_pos = item_loc.get_inventory_position();
+
+    if ( inv_pos != INT_MIN ) {
+        u.reading_from_inventory = true;
+        draw();
+        u.read( inv_pos );
         return;
     }
-    draw();
-    u.read(pos);
+
+    u.reading_from_inventory = false;
+
+    item *it = item_loc.get_item();
+    if ( it == nullptr ) {
+        add_msg( _("Never mind.") );
+        return;
+    }
+
+
+    // check if in vehicle as item cannot be in inventory since the above check "inv_pos != INT_MIN" already handle item in inventory
+    // at the moment, only item_location::item_on_person() will set values returned for get_inventory_position(). All others will return INT_MIN
+    bool item_on_veh = false;
+
+    int part = -1;
+    int item_index = 0;
+    vehicle *veh = m.veh_at( u.pos(), part );
+    point veh_pt( INT_MIN, INT_MIN );
```

Comment:
Wow, thanks BevapDin,

Did not know about player::can_pickup.
Will look into it after figuring out the player_activity. It is quite complicated for me.

Haha, at first thought of taking up this issue as it seems to be simplest for a newbie.
Seems like it is far more complicated than it looks :)